File: ruby/linked-list/linked_list.rb
class Deque
  private

  attr_accessor :head, :tail

  class Node
    attr_reader :value
    attr_accessor :next, :prev

    def initialize(value)
      @value = value
    end
  end

  public

  def size
    count = 0
    node = head
    until node.nil?
      count += 1
      node = node.next
    end
    count
  end

  def push(value)
    node = Node.new(value)
    if tail.nil?
      self.head = node
    else
      node.prev = tail
      tail.next = node
    end
    self.tail = node
    self
  end

  def pop
    node = tail
    return if node.nil?

    if node.prev.nil?
      self.tail = self.head = nil
    else
      node.prev.next = nil
      self.tail = node.prev
    end
    node.value
  end

  def unshift(value)
    node = Node.new(value)
    if head.nil?
      self.tail = node
    else
      node.next = head
      head.prev = node
    end
    self.head = node
    self
  end

  def shift
    node = head
    return if node.nil?

    if node.next.nil?
      self.tail = self.head = nil
    else
      node.next.prev = nil
      self.head = node.next
    end
    node.value
  end

  def delete(value)
    node = head
    until node.nil?
      if node.value == value
        remove(node)
        break
      end
      node = node.next
    end
  end

  private

  def remove(node)
    if node.next.nil?
      self.tail = node.prev
    else
      node.next.prev = node.prev
    end

    if node.prev.nil?
      self.head = node.next
    else
      node.prev.next = node.next
    end
  end
end

File: go/linked-list/linked_list.go
package linkedlist

// TODO when checking for empty list and pop/unshift
// import "errors"

type Node struct {
	Value interface{}
	next  *Node
	prev  *Node
}

type List struct {
	head *Node
	tail *Node
}

func NewList(args ...interface{}) *List {
	lst := List{}
	for _, val := range args {
		lst.Push(val)
	}
	return &lst
}

func (n *Node) Next() *Node {
	return n.next
}

func (n *Node) Prev() *Node {
	return n.prev
}

func (l *List) Unshift(v interface{}) {
	n := Node{Value: v}
	if l.head == nil {
		l.head = &n
		l.tail = &n
	} else {
		l.head.prev = &n
		n.next = l.head
		l.head = &n
	}
}

func (l *List) Shift() (interface{}, error) {
	// TODO check for empty list
	n := l.head
	if n.next == nil {
		l.head = nil
		l.tail = nil
	} else {
		l.head = n.next
		l.head.prev = nil
		n.next = nil // helpful for garbage collection?
	}
	return n.Value, nil
}

func (l *List) Push(v interface{}) {
	n := Node{Value: v}
	if l.tail == nil {
		l.head = &n
		l.tail = &n
	} else {
		l.tail.next = &n
		n.prev = l.tail
		l.tail = &n
	}
}

func (l *List) Pop() (interface{}, error) {
	// TODO check for empty list
	n := l.tail
	if n.prev == nil {
		l.head = nil
		l.tail = nil
	} else {
		l.tail = n.prev
		l.tail.next = nil
		n.prev = nil // helpful for garbage collection?
	}
	return n.Value, nil
}

func (l *List) Reverse() {
	newList := List{}
	for n := l.head; n != nil; n = n.next {
		newList.Unshift(n.Value)
	}
	l.head = newList.head
	l.tail = newList.tail
}

func (l *List) First() *Node {
	return l.head
}

func (l *List) Last() *Node {
	return l.tail
}
